### PR TITLE
Make the bigtable::ClientAdmin and bigtable::DataAdmin interfaces thread-safe

### PR DESCRIPTION
**Pull request by coryan**
This is part of the changes for #101.  Finally we change the interface to return a type that can be made thread safe.  All is left after this is a bit of refactoring and adding some mutexes.

### CODE REVIEW

**Comment by coryan** [comment]
Ping.

File: bigtable/admin/admin_client.cc [hunk 1]
```diff
@@ -36,20 +36,40 @@ class SimpleAdminClient : public bigtable::AdminClient {
       : project_(std::move(project)), options_(std::move(options)) {}
 
   std::string const& project() const override { return project_; }
-  void on_completion(grpc::Status const& status) override;
-  ::google::bigtable::admin::v2::BigtableTableAdmin::StubInterface&
-  table_admin() override;
+  void on_completion(grpc::Status const& status) override {
+    if (status.ok()) {
```
[the remark this answers is not in the document]
Comment (coryan) [reply bigtable/admin/admin_client.cc]: Done.

**Comment by coryan** [comment]
PTAL.

**Comment by mbrukman** [comment]
Merging this PR seems to have broken the build on `master`; see [the build](https://travis-ci.org/GoogleCloudPlatform/google-cloud-cpp/builds/325287604), specifically [this job](https://travis-ci.org/GoogleCloudPlatform/google-cloud-cpp/jobs/325287611).

The failure appears to be in the step to build `cbt` via `go get`:

```
Step 5/7 : ARG GOPATH=/var/tmp/build/cbt
 ---> Running in 41b100c93352
Removing intermediate container 41b100c93352
 ---> 1a5c626496c9
Step 6/7 : WORKDIR ${GOPATH}
Removing intermediate container 363a8e4ac407
 ---> aa811117610f
Step 7/7 : RUN go get cloud.google.com/go/bigtable/cmd/cbt     cloud.google.com/go/bigtable/cmd/emulator
 ---> Running in 6f964f4857d5
# cd .; git clone https://code.googlesource.com/google-api-go-client /var/tmp/build/cbt/src/google.golang.org/api
Cloning into '/var/tmp/build/cbt/src/google.golang.org/api'...
fatal: remote error: Git repository not found
package google.golang.org/api/option: exit status 128
package google.golang.org/api/iterator: cannot find package "google.golang.org/api/iterator" in any of:
	/usr/lib/golang/src/google.golang.org/api/iterator (from $GOROOT)
	/var/tmp/build/cbt/src/google.golang.org/api/iterator (from $GOPATH)
package google.golang.org/api/transport: cannot find package "google.golang.org/api/transport" in any of:
	/usr/lib/golang/src/google.golang.org/api/transport (from $GOROOT)
	/var/tmp/build/cbt/src/google.golang.org/api/transport (from $GOPATH)
package google.golang.org/api/cloudresourcemanager/v1: cannot find package "google.golang.org/api/cloudresourcemanager/v1" in any of:
	/usr/lib/golang/src/google.golang.org/api/cloudresourcemanager/v1 (from $GOROOT)
	/var/tmp/build/cbt/src/google.golang.org/api/cloudresourcemanager/v1 (from $GOPATH)
package google.golang.org/api/transport/grpc: cannot find package "google.golang.org/api/transport/grpc" in any of:
	/usr/lib/golang/src/google.golang.org/api/transport/grpc (from $GOROOT)
	/var/tmp/build/cbt/src/google.golang.org/api/transport/grpc (from $GOPATH)
The command '/bin/sh -c go get cloud.google.com/go/bigtable/cmd/cbt     cloud.google.com/go/bigtable/cmd/emulator' returned a non-zero code: 1
```

This may be transient, or it might not. I restarted that specific job to see if it will pass on a second try (the URL it's complaining about does exist); if not, this may need to be revisited.

**Comment by coryan** [comment]
That is an unfortunately common occurrence with `go get`.  I have restarted multiple builds because of it.  Let me put together a PR to reduce the issue.

**Comment by coryan** [comment]
Ahh, no, this is something else.  We are living at head, and head is broken.

**Comment by mbrukman** [comment]
In that case, maybe we should be getting `cbt` via `gcloud` components? That always ships with a pre-built, tested version of `cbt`.

Per https://cloud.google.com/bigtable/docs/go/cbt-overview, once `gcloud` is installed:

```
gcloud components update
gcloud components install cbt
```

**Comment by garye** [comment]
Whatever we do here would be good to document as a supported way to use the emulator for travis or other CI builds. go get/gcloud/docker are options here.

**Pull request merged**